### PR TITLE
Add push command

### PR DESCRIPTION
**Pull request by sema**
First draft of adding a push command. PR not complete yet (missing some logic and tests). Getting this out there to gather some early feedback.

### CODE REVIEW

File: lib/snapshot/utils.go [hunk 1]
```diff
@@ -178,6 +178,37 @@ func CreateTarFromDirectory(target, dir string) error {
 	})
 }
 
+// CreateDirectoryFromTar creates a directory containing the contents of the given
+// tar archive.
+func CreateDirectoryFromTar(target, tar string) error {
```
Comment (yiranwang52) [inline lib/snapshot/utils.go]: this function should be in tario package.
Initiate `NewGzipReader`, use `.Next()` to iterate through items like in [UpdateFromTarReader](https://github.com/uber/makisu/blob/c75663b197c972ff65999db79c36eae6f622b91d/lib/snapshot/mem_fs.go#L156). No need to care too much about actual content.

File: lib/docker/cli/image.go [hunk 2]
```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
 	return os.Open(targetPath)
 }
 
+// WriteTar imports an image, as a tar, to the image store
+func (tarer DefaultImageTarer) WriteTar(imageName image.Name, tarPath string) error {
```
Comment (yiranwang52) [inline lib/docker/cli/image.go]: `ImportTar` is better

File: lib/docker/cli/image.go [hunk 3]
```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
...
+		configPath := filepath.Join(dir, exportManifest.Config.String())
+		configID := exportManifest.Config.ID()
+		if err = tarer.store.Layers.LinkStoreFileFrom(configID, configPath); err != nil && !os.IsExist(err) {
+			return fmt.Errorf("commit config to store: %s", err)
+		}
+
+		// Import extracted dir content into image store -- {sha}/layer.tar
+		for _, layer := range exportManifest.Layers {
+			layerPath := path.Join(dir, layer.String())
+			layerID := layer.ID()
```
Comment (yiranwang52) [inline lib/docker/cli/image.go]: if the export manifest is from `docker save`, the ID/SHA would actually be random (though the ordering would be correct). Need to fix that.

File: lib/docker/cli/image.go [hunk 4]
```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
...
+		}
+		if _, err := distManifestFile.Write(distManifestJSON); err != nil {
+			return fmt.Errorf("write manifest file: %s", err)
+		}
+		if err := distManifestFile.Close(); err != nil {
+			return fmt.Errorf("close manifest file: %s", err)
+		}
+
+		distManifestPath := distManifestFile.Name()
+		if err = tarer.store.Manifests.LinkStoreFileFrom(repo, tag, distManifestPath); err != nil && !os.IsExist(err) {
```
Comment (yiranwang52) [inline lib/docker/cli/image.go]: err :=

File: lib/docker/cli/image.go [hunk 5]
```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
...
+			return fmt.Errorf("commit manifest to store: %s", err)
+		}
+
+		// Import extracted dir content into image store -- {sha}.json
+		configPath := filepath.Join(dir, exportManifest.Config.String())
+		configID := exportManifest.Config.ID()
+		if err = tarer.store.Layers.LinkStoreFileFrom(configID, configPath); err != nil && !os.IsExist(err) {
```
Comment (yiranwang52) [inline lib/docker/cli/image.go]: err :=

File: lib/docker/cli/image.go [hunk 6]
```diff
@@ -93,6 +88,79 @@ func (tarer DefaultImageTarer) CreateTarReader(imageName image.Name) (io.Reader,
...
+			return fmt.Errorf("commit config to store: %s", err)
+		}
+
+		// Import extracted dir content into image store -- {sha}/layer.tar
+		for _, layer := range exportManifest.Layers {
+			layerPath := path.Join(dir, layer.String())
+			layerID := layer.ID()
+			if err = tarer.store.Layers.LinkStoreFileFrom(layerID, layerPath); err != nil && !os.IsExist(err) {
```
Comment (yiranwang52) [inline lib/docker/cli/image.go]: err :=

File: test/python/test_push.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import subprocess
+import tempfile
+import utils
+
+import image
+
+
+def test_push_simple(registry1, registry2, storage_dir, tmpdir):
+    new_image = utils.new_image_name()
+    # replica_image = utils.new_image_name()
```
Comment (sema) [inline test/python/test_push.py]: rm replica related lines - let us add those in a separate test when we fix replica support

File: bin/makisu/cmd/push.go [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+		if err := cmd.pushImage(store, target); err != nil {
+			return fmt.Errorf("failed to push image: %s", err)
+		}
+	}
+	for _, replica := range cmd.replicas {
+		target := image.MustParseName(replica)
+		if err := cmd.pushImage(store, target); err != nil {
+			return fmt.Errorf("failed to push image: %s", err)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yiranwang52) [reply bin/makisu/cmd/push.go]: yes.

File: test/python/test_push.py (hunk 7 above)
Comment (yiranwang52) [reply test/python/test_push.py]: okay

**Pull request merged**